File: 565 Divisibility of sum of divisors.pl
#!/usr/bin/perl

# Daniel "Trizen" Șuteu
# Date: 04 July 2018
# https://github.com/trizen

# https://projecteuler.net/problem=565

# Runtime: ~4 minutes

use 5.010;
use strict;
use warnings;

use ntheory qw(is_prime divisor_sum logint sqrtint forprimes);

sub S {
    my ($n, $k) = @_;

    my %seen;
    my $sum = 0;

    my $process_term = sub {
        my ($t) = @_;

        for (my $s = $t ; $s <= $n ; $s += $t) {

            next if ($s % $t**2 == 0);

            if (divisor_sum($s / $t) % $k == 0) {
                next if (++$seen{$s} >= 2);
            }

            if (divisor_sum($s) % $k == 0) {
                $sum += $s;
            }
        }
    };

    forprimes {
        my $p = $_;
        foreach my $e (3 .. 1 + logint($n, $p)) {
            if (($p**$e - 1) % $k == 0) {
                $process_term->($p**($e - 1));
            }
        }
    } sqrtint($n);

    for (my $t = $k ; $t <= $n ; $t += $k) {
        if (is_prime($t - 1)) {
            $process_term->($t - 1);
        }
    }

    return $sum;
}

say S(1e6,  2017);
say S(1e9,  2017);
say S(1e11, 2017);

File: Julia/565 Divisibility of sum of divisors.jl
#!/usr/bin/julia

# Daniel "Trizen" Șuteu
# Date: 04 July 2018
# https://github.com/trizen

# https://projecteuler.net/problem=565

# Runtime: ~18 minutes

using Primes

function divisor_sum(n)
    sigma = 1
    for (p,e) in factor(n)
        s = 1
        for j in 1:e
            s += p^j
        end
        sigma *= s
    end
    sigma
end

function S(n::Int64, k::Int64)

    total = 0
    seen = Dict{Int64,Bool}()

    function process_term(t::Int64)

        for s in t:t:n

            if (s % (t*t) == 0)
                continue
            end

            if (divisor_sum(div(s,t)) % k == 0)
                if (haskey(seen, s))
                    continue
                else
                    seen[s] = true
                end
            end

            if (divisor_sum(s) % k == 0)
                total += s
            end
        end
    end

    p = 2
    isqrtn = isqrt(n)

    while (p <= isqrtn)

        for e in 3 : 1+Int64(floor(log(p, n)))
            if ((p^e - 1) % k == 0)
                process_term(p^(e-1))
            end
        end

        p = nextprime(p+1)
    end

    for t in k:k:n
        if (isprime(t - 1))
            process_term(t - 1)
        end
    end

    total
end

println(S(10^6,  2017))
println(S(10^9,  2017))
println(S(10^11, 2017))
